### PR TITLE
mildwonkey/b-show-state

### LINKED ISSUE #20028: terraform show should show config
Hi All,

I'm expecting the terraform config to appear in the json output. I need the instance type and potentially also the ami, but I'm not seeing it. Thanks!

### Terraform Version

4e14ab755758c312c8b4664722c0ac4aefcbb005

### Terraform Configuration Files
```hcl
resource "aws_instance" "basic" {
  ami = "ami-0a16c82eb0cc3aad8"
  instance_type = "t2.micro"
  key_name = "aws-test-west"
}

provider "aws" {
    region = "us-west-1"
}
```

### Actual Output
```json
{
  "format_version": "0.1",
  "planned_values": {
    "root_module": {
      "resources": [
        {
          "address": "aws_instance.basic",
          "mode": "managed",
          "type": "aws_instance",
          "name": "basic",
          "provider_name": "aws",
          "schema_version": 1
        }
      ]
    }
  },
  "resource_changes": [
    {
      "address": "aws_instance.basic",
      "mode": "managed",
      "type": "aws_instance",
      "name": "basic",
      "deposed": true,
      "change": {
        "actions": [
          "Create"
        ],
        "before": null
      }
    }
  ],
  "configuration": {
    "provider_config": {
      "aws": {
        "name": "aws",
        "expressions": {
          "region": {
            "constant_value": "us-west-1"
          }
        }
      }
    },
    "root_module": {
      "resources": [
        {
          "address": "aws_instance.basic",
          "mode": "managed",
          "type": "aws_instance",
          "name": "basic",
          "provider_config_key": "provider.aws",
          "schema_version": 0,
          "count_expression": {},
          "for_each_expression": {}
        }
      ]
    }
  }
}
```

### Steps to Reproduce
1. terraform init
2. terraform plan -out terraform.tfplan
3. terraform show -json terraform.tfplan

### PR DESCRIPTION
**Pull request by mildwonkey**
Various bug fixes:
* attribute values are now marshaled to json.RawMessages using ctyjson. Before this they were being passed a cty.Values and lost entirely when marshaled to json
* planned_values root module is back! There was some flawed logic where planned values was only showing changes, not no-ops. (deleted values are not seen in planned_values)
* fixing awkward (and wrong) test name.
* properly list change actions e.g. "change", "delete" instead of using stringified plans.Action
* use unique provider keys in jsonconfig so aliased providers don't get merged together

fixes #20028 (sort of - this PR is coming before support for showing state only is merged)

### CODE REVIEW

**Comment by mwhooker** [comment]
Working off this branch, I'm not seeing the `after` object in the plan json output. This is working from an empty state to creating 1 aws json instance

```
{
  "format_version": "0.1",
  "planned_values": {
    "root_module": {
      "resources": [
        {
          "address": "aws_instance.basic",
          "mode": "managed",
          "type": "aws_instance",
          "name": "basic",
          "provider_name": "aws",
          "schema_version": 1
        }
      ]
    }
  },
  "resource_changes": [
    {
      "address": "aws_instance.basic",
      "mode": "managed",
      "type": "aws_instance",
      "name": "basic",
      "deposed": true,
      "change": {
        "actions": [
          "Create"
        ],
        "before": null
      }
    }
  ],
  "configuration": {
    "provider_config": {
      "aws": {
        "name": "aws",
        "expressions": {
          "region": {
            "constant_value": "us-west-1"
          }
        }
      }
    },
    "root_module": {
      "resources": [
        {
          "address": "aws_instance.basic",
          "mode": "managed",
          "type": "aws_instance",
          "name": "basic",
          "provider_config_key": "provider.aws",
          "schema_version": 0,
          "count_expression": {},
          "for_each_expression": {}
        }
      ]
    }
  }
}
```

**Comment by mildwonkey** [comment]
🤔  We're not seeing anything in the `after` for the aws_instance because the values are mixed known and unknown:

https://github.com/hashicorp/terraform/blob/5adc9187134d4cc797b259e720a33b920286b24e/command/jsonplan/values.go#L64

I will ask around and see what my options are.

**Comment by mildwonkey** [comment]
@mwhooker I think I've got it now!

File: command/jsonplan/plan.go [hunk 1]
```diff
@@ -158,31 +159,40 @@ func (p *plan) marshalResourceChanges(changes *plans.Changes, schemas *terraform
...
+						return cty.False, nil
+					}
+
+					if !val.Type().IsPrimitiveType() {
+						return val, nil // just pass through non-primitives; they already contain our transform results
+					}
+
+					if val.IsKnown() {
+						// null rather than false here so that known values
+						// don't appear at all in JSON serialization of our result
```
Comment (apparentlymart) [inline command/jsonplan/plan.go]: This comment is no longer accurate

Comment (mildwonkey) [reply command/jsonplan/plan.go]: whoops, thank you.

**Comment by mwhooker** [comment]
yes, this looks beautiful, thank you! I think I have all the info I need now

**Pull request merged**